### PR TITLE
Better stable rearrange sampling

### PR DESCRIPTION
**Pull request by aclegg3**
## Motivation and Context
This PR adjusts some generator parameters for object placement sampling to enable more robust performance and faster generation times on HSSD.

Primarily this: 1. samples object COM placements about the Receptacle such that raycasts don't begin inside of convex shapes and miss surfaces. 2. better accounts for object margins causing similar issues.

## How Has This Been Tested

Tested with rearrange generator on HSSD scenes. MUCH better performance that `main`. Episodes generated in 1-2 attempts instead of 10-20 attempts.

Locally with [habitat-sim 2078](https://github.com/facebookresearch/habitat-sim/pull/2078).

## Types of changes

- **\[Development\]** A pull request that add new features to the [habitat-lab](/habitat-lab) task and environment codebase. Development Pull Requests must be small (less that 500 lines of code change), have unit testing, very extensive documentation and examples. These are typically new tasks, environments, sensors, etc... The review process for these Pull Request is longer because these changes will be maintained by our core team of developers, so make sure your changes are easy to understand!

## Checklist

- [x] My code follows the code style of this project.
- [x] I have updated the documentation if required.
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes if required.

### CODE REVIEW

File: habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py [hunk 1]
```diff
@@ -137,6 +138,8 @@ def bb_ray_prescreen(
     obj: habitat_sim.physics.ManagedRigidObject,
     support_obj_ids: Optional[List[int]] = None,
     check_all_corners: bool = False,
+    estimate_support_stability: bool = True,
```
Comment (aclegg3) [inline habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py:141]: NOTE:
This seems to help in edge cases (as shown in the videos) with very high sample numbers.
However, in practice with typical scenarios it doesn't do much and actually leads to slightly worse generator performance.

File: habitat-lab/habitat/datasets/rearrange/run_episode_generator.py [hunk 2]
```diff
@@ -4,12 +4,17 @@
...
 
+import ctypes
 import os
 import os.path as osp
 import random
+import sys
 from dataclasses import dataclass, field
 from typing import TYPE_CHECKING, Any, List
 
+flags = sys.getdlopenflags()
```
Comment (Skylion007) [inline habitat-lab/habitat/datasets/rearrange/run_episode_generator.py:15]: We really should have a global entry point hat ensures this gets gets triggered.

File: habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py [hunk 3]
```diff
@@ -153,7 +156,7 @@ def bb_ray_prescreen(
     lowest_key_point_height = None
     highest_support_impact: mn.Vector3 = None
```
Comment (Skylion007) [inline habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py:157]:
```suggestion
    highest_support_impact: Optional[mn.Vector3]
```

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py [hunk 4]
```diff
@@ -800,10 +810,49 @@ def __init__(
...
+                for filter_type in [
+                    "manually_filtered",
+                    "access_filtered",
+                    "stability_filtered",
+                    "height_filtered",
+                ]:
+                    for filtered_unique_name in filter_json[filter_type]:
+                        filtered_unique_names.append(filtered_unique_name)
+            # add exclusion filters to all receptacles sets
+            for _, r_set in self._receptacle_sets.items():
```
Comment (Skylion007) [inline habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:844]:
```suggestion
            for r_set in self._receptacle_sets.values():
```

File: habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py (hunk 3 above)
Comment (Skylion007) [reply habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py]: Looks like we need to import Optional from typing.

File: habitat-lab/habitat/datasets/rearrange/samplers/object_sampler.py [hunk 5]
```diff
@@ -39,6 +39,7 @@ def __init__(
         sample_region_ratio: Optional[Dict[str, float]] = None,
         nav_to_min_distance: float = -1.0,
         recep_set_sample_probs: Optional[Dict[str, float]] = None,
+        translation_up_offset=0.08,
```
Comment (Skylion007) [inline habitat-lab/habitat/datasets/rearrange/samplers/object_sampler.py:42]:
```suggestion
        translation_up_offset : float =0.08,
```

**Review by Skylion007 — APPROVED** [review]

**Pull request merged**